support/k8s/k8s-workload-registrar/controllers/pod_controller.go

```diff
@@ -0,0 +1,313 @@
+		if selector.Type == "k8s" {
+			splitted := strings.SplitN(selector.Value, ":", 2)
+			if len(splitted) > 1 {
+				switch WorkloadSelectorSubType(splitted[0]) {
+				case PodNamespaceSelector:
+					podNamespace = splitted[1]
+				case PodNameSelector:
+					podName = splitted[1]
+				}
+			}
+		}
+	}
+	if podNamespace != "" && podName != "" {
+		return &types.NamespacedName{
+			Namespace: podNamespace,
+			Name:      podName,
+		}
+	}
+	return nil
+}
+
+func (r *PodReconciler) makeSpiffeID(obj ObjectWithMetadata) string {
+	return r.makeSpiffeIDForPod(obj.(*corev1.Pod))
+}
+
+func (r *PodReconciler) mungeIP(ip string) string {
+	if strings.Contains(ip, ".") {
+		// IPv4
+		return strings.Replace(ip, ".", "-", -1)
+	}
+	if strings.Contains(ip, ":") {
+		// IPv6
+		return strings.Replace(ip, ":", "-", -1)
+	}
+	return ip
+}
+
+func (r *PodReconciler) fillEntryForPod(ctx context.Context, entry *common.RegistrationEntry, pod *corev1.Pod) (*common.RegistrationEntry, error) {
+	if !r.AddPodDNSNames {
+		return entry, nil
+	}
+
+	names := make(map[string]bool)
+
+	endpointsList := corev1.EndpointsList{}
+	if err := r.List(ctx, &endpointsList, client.InNamespace(pod.Namespace), client.MatchingFields{endpointSubsetAddressReferenceField: pod.Name}); err != nil {
+		return nil, err
+	}
+
+	for _, endpoints := range endpointsList.Items {
+		endpoints := endpoints
+		// Based on https://github.com/kubernetes/dns/blob/master/docs/specification.md
+		// We cheat slightly and don't check the type of service (headless or not), we just add all possible names.
+
+		// 2.3.1 and 2.4.1: <service>.<ns>.svc.<zone>
+		names[fmt.Sprintf("%s.%s.svc.%s", endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
+
+		r.forEachPodEndpointAddress(&endpoints, func(address corev1.EndpointAddress) {
+			if pod.Name == address.TargetRef.Name && pod.Namespace == address.TargetRef.Namespace {
+				// 2.4.1: <hostname>.<service>.<ns>.svc.<zone>
+				if address.Hostname != "" {
+					names[fmt.Sprintf("%s.%s.%s.svc.%s", address.Hostname, endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
+				} else {
+					// The spec leaves this case up to the implementation, so here we copy CoreDns...
+					// CoreDNS has an option to switch between the following two options. We don't have that flag, so
+					// we'll just add both pod name and IP based name for now.
+					names[fmt.Sprintf("%s.%s.%s.svc.%s", address.TargetRef.Name, endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
+					names[fmt.Sprintf("%s.%s.%s.svc.%s", r.mungeIP(address.IP), endpoints.Name, endpoints.Namespace, r.ClusterDNSZone)] = true
+				}
+			}
+		})
+	}
+
+	// TODO: The docs at https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/#pods claim
+	// pod-ip-address.deployment-name.my-namespace.svc.cluster-domain.example too.
+	// This behaviour isn't in the k8s DNS spec, and I can't see how CoreDNS implement this, so we'll ignore it for now.
```

Comment:
Not looked into in detail, but maybe this is where it happens: https://github.com/coredns/coredns/blob/9230e1772bc24faddd94a3dc9be7fe52a4d29e83/plugin/kubernetes/kubernetes.go#L280-L294

Reply:
I'm pretty convinced that `pod-ip-address.deployment-name.my-namespace.svc.cluster-domain.example` doesn't actually exist, but I did miss `pod-ip-address.my-namespace.pod.cluster-domain.example`. I've corrected that omission, updated the comments, and added some tests.